hack/generator/spec/versioning.md

```diff
@@ -0,0 +1,201 @@
+
+## Non-Goals
+
+**Coverage of every case by code generation:** While it's likely that very high coverage will be achievable with code generation, we don't believe that it will be practical to handle every possible situation automatically. It's therefore necessary for the solution to have some form of extensibility allowing for the injection of hand written code.
+
+## Proposed Solution
+
+In summary (details below):
+
+* Base the storage version on the latest available API version for each resource type
+* Augment the storage version with a name/value property bag for storing additional information
+* Generate the required `ConvertTo()` and `ConvertFrom()` methods automatically
+* Allow configuration to support common changes from version to version
+* Define extension points for hand written conversions
+
+### Use the latest available API Version
+
+As a starting point for defining the storage schema for each resource type, we will use the latest available version of the matching ARM API. This ensures maximum compatibility with that API when it is used, giving a good starting point for new adoption of the service operator.
+
+If a resource type has been dropped from the ARM API, we will still generate a storage schema for it based on the last ARM API version where it existed; this ensures backward compatibility with existing service operator deployments.
+
+### Augment storage with a name/value property bag
+
+To facilitate easy and safe storage of additional configuration in storage, we'll include a name/value property bag for storage, along with helper methods that allow for strongly typed access to values stored within.
+
+Sequestering the information away within the storage schema is more robust than using separate annotations as they are less subject to arbitary modification by users. This allows us to largely avoid situations where well meaning (but uninformed) consumers of the service operator innocently make changes that result in the operator becoming failing. We particularly want to avoid this failure mode because recovery will be difficult - restoration of the modified/deleted information may be impractical or impossible.
+
+---
+
+****Outstanding issue:** Where do the helper functions for the name/value property bag reside?**
+
+We could generate those anew for each type, making them conveniently available and avoiding the need for our generated types to depend on anything more than the standard Go library Alternatively, we could define the property bag as a new struct, wrapping up storage & helper methods into a reusable form.
+
+---
+
+### Generate conversion methods
+
+Each of the structs generated for ARM API will have the normal `ConvertTo()` and `ConvertFrom()` methods generated automatically, mapping like-for-like properties between the hub version and the current type.
+
+* When the name and type of properties match exactly, conversion in both directions will be simple assignments.
```

Comment:
Yeah, you have to deepcopy everything...

Reply:
I definitely need to tweak the wording here as that's exactly what I thought I was saying! If both types have the same property, we can just generate an assignment to copy the data:

```
package vStorage
type Person struct {
    Name string
}

package v20200601
type Person struct {
    Name string
}

func (person *Person) ConvertTo() vStorage.Person {
    result := &vStorage.Person{}

    result.Name = person.Name

    return result
}
```

_Indicative syntax only_.

Reply:
yeah but that's the simplest case - holding true for structs comprised of simple (basic type) properties only.

If you imagine a more complicated struct like:
```
package vStorage
type Person struct {
    Name string
}

type School {
    Students []*Person
}

package v20200601
type Person struct {
    Name string
}

type School {
    Students []*Person
}

func (school *School) ConvertTo() vStorage.School {
    result := &vStorage.School{}

    // This is not just a basic assignment
    for _, person := school.Students {
        result.Students = append(result.Students, person.CovertTo())
    }
    return result
}
```

I'm not super concerned about the specifics of all the different conversions because I'm convinced we can do them all, I'm just highlighting that _a lot of the time_ it's not going to be a simple assignment because many objects in Azure are quite deeply nested, and we need to generate conversion for structs, maps, slices as well. Also (from this), _every single type_ in a version must be convertible to the corresponding type in the storage version if there is one (so there are going to be a lot of `ConvertTo`'s flying around)

Reply:
Yes, there will be a large number of conversion methods in flight.

Reply:
Reworked this section.